### ISSUE TITLE
Vulnerability roundup 111: libspf2-1.2.10: 2 advisories [9.8]

### ISSUE THREAD

**Issue by ckauhaus** [NixOS/nixpkgs#160671]
[search](https://search.nix.gsc.io/?q=libspf2&i=fosho&repos=NixOS-nixpkgs), [files](https://github.com/NixOS/nixpkgs/search?utf8=%E2%9C%93&q=libspf2+in%3Apath&type=Code)

* [ ] [CVE-2021-33912](https://nvd.nist.gov/vuln/detail/CVE-2021-33912) CVSSv3=9.8 (nixos-21.11)
* [ ] [CVE-2021-33913](https://nvd.nist.gov/vuln/detail/CVE-2021-33913) CVSSv3=9.8 (nixos-21.11)

### CVE-2021-33912

libspf2 before 1.2.11 has a four-byte heap-based buffer overflow that might allow remote attackers to execute arbitrary code (via an unauthenticated e-mail message from anywhere on the Internet) with a crafted SPF DNS record, because of incorrect sprintf usage in SPF_record_expand_data in spf_expand.c. The vulnerable code may be part of the supply chain of a site's e-mail infrastructure (e.g., with additional configuration, Exim can use libspf2; the Postfix web site links to unofficial patches for use of libspf2 with Postfix; older versions of spfquery relied on libspf2) but most often is not.

### CVE-2021-33913

libspf2 before 1.2.11 has a heap-based buffer overflow that might allow remote attackers to execute arbitrary code (via an unauthenticated e-mail message from anywhere on the Internet) with a crafted SPF DNS record, because of SPF_record_expand_data in spf_expand.c. The amount of overflowed data depends on the relationship between the length of an entire domain name and the length of its leftmost label. The vulnerable code may be part of the supply chain of a site's e-mail infrastructure (e.g., with additional configuration, Exim can use libspf2; the Postfix web site links to unofficial patches for use of libspf2 with Postfix; older versions of spfquery relied on libspf2) but most often is not.

-----
Scanned versions: nixos-21.11: 2128d0aa28e.

Cc @pacien

**Comment by pacien** [comment]
Following issue #138683, we have switched to the more up-to-date fork
helsinki-systems/libspf2 (2.2.12) on nixpkgs/master, but we are still using
shevek/libspf2 (1.2.10) on nixpkgs/release-21.11.

I just checked: the fork seems to have the fixes for the two CVEs listed in the
current issue, plus other fixes buffers/use-after-free fixes.

Therefore, I think we should backport PR #149589 to release-21.11 as well.
